drivers/dma/dma_npcx_gdma.c

```diff
@@ -0,0 +1,458 @@
+/*
+ * Copyright (c) 2025 Nuvoton Technology Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <soc.h>
+#include <string.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/dma.h>
+#include <zephyr/init.h>
+#include <zephyr/irq.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/__assert.h>
+
+#define DT_DRV_COMPAT nuvoton_npcx_gdma
+
+LOG_MODULE_REGISTER(dma_npcx, CONFIG_DMA_LOG_LEVEL);
+
+enum {
+	DMA_NPCX_GDMAMS_SOFT = 0,
+	DMA_NPCX_GDMAMS_REQ0,
+	DMA_NPCX_GDMAMS_REQ1,
+};
+
+enum {
+	DMA_NPCX_GDMA_TWS_1B = 0,
+	DMA_NPCX_GDMA_TWS_2B,
+	DMA_NPCX_GDMA_TWS_4B,
+};
+
+enum {
+	DMA_NPCX_BURST_LEN_1B  = 1, /* Burst length is 1 byte */
+	DMA_NPCX_BURST_LEN_2B  = 2, /* Burst length is 2 bytes */
+	DMA_NPCX_BURST_LEN_4B  = 4,  /* Burst length is 4 bytes */
+	DMA_NPCX_BURST_LEN_16B = 16  /* Burst length is 16 bytes */
+};
+
+struct dma_npcx_ch_data {
+	int channel_active; /* busy is 1 and idle is 0 */
+	uint32_t channel_dir;
+	struct k_mutex ch_mutex;
+	void *user_data;
+	dma_callback_t callback; /* GDMA transfer finish callback function */
+};
+
+/* Device config */
+struct dma_npcx_config {
+	/* DMA controller base address */
+	struct gdma_reg **reg_base;
+	uint8_t max_channel;
+	struct npcx_clk_cfg clk_cfg;
+	void (*irq_config)(void);
+};
+
+struct dma_npcx_dev_data {
+	struct dma_npcx_ch_data *channels;
+};
+
+/* Driver convenience defines */
+#define HAL_INSTANCE(dev, channel)                                                                 \
+	((struct gdma_reg *)(((const struct dma_npcx_config *)(dev)->config)->reg_base[channel]))
+
+static void dma_set_power_down(const struct device *dev, uint8_t channel, bool enable)
+{
+	struct gdma_reg *const inst = HAL_INSTANCE(dev, channel);
+
+	if (enable) {
+		inst->CONTROL |= BIT(NPCX_DMACTL_GPD);
+	} else {
+		inst->CONTROL &= ~BIT(NPCX_DMACTL_GPD);
+	}
+}
+
+static void dma_npcx_isr(const struct device *dev)
+{
+	const struct dma_npcx_config *const dev_cfg = dev->config;
+	struct dma_npcx_dev_data *const dev_data = dev->data;
+	struct dma_npcx_ch_data *channel_data;
+
+	for (int channel = 0; channel < dev_cfg->max_channel; channel++) {
+		struct gdma_reg *const inst = HAL_INSTANCE(dev, channel);
+
+		channel_data = &dev_data->channels[channel];
+#if defined(CONFIG_DMA_NPCX_GDMA_EX)
+		/* Check GDMA Transfer Error */
+		if (IS_BIT_SET(inst->CONTROL, NPCX_DMACTL_GDMAERR)) {
+			LOG_ERR("GDMA transfer error occurred!");
+		}
+#endif
+		/* Clear GDMA interrupt flag */
+		if (IS_BIT_SET(inst->CONTROL, NPCX_DMACTL_TC)) {
+			inst->CONTROL &= ~BIT(NPCX_DMACTL_TC);
+		} else {
+			continue;
+		}
```

Comment:
This should look more readable IMO, grouping clearing the bit with handling the interrupt

```suggestion
		if (!IS_BIT_SET(inst->CONTROL, NPCX_DMACTL_TC)) {
			continue;
		}

		inst->CONTROL &= ~BIT(NPCX_DMACTL_TC);
```

Reply:
@TomChang19 thank you for applying my review suggestions! Feel free to resolve the threads if you want

Reply:
Done. Thanks for your suggestions.